epsagon/aws_sdk_v2_factories/s3_factory.go

```diff
@@ -28,76 +27,92 @@ func S3EventDataFactory(
 	handleSpecificOperation(r, res, metadataOnly, handleSpecificOperations, nil, currentTracer)
 }
 
-func commonS3OpertionHandler(r *aws.Request, res *protocol.Resource, metadataOnly bool) {
-	inputValue := reflect.ValueOf(r.Params).Elem()
+func commonS3OperationHandler(r *AWSCall, res *protocol.Resource, metadataOnly bool) {
+	responseValue := reflect.ValueOf(r.Res).Elem()
+	inputValue := reflect.ValueOf(r.Input).Elem()
+	outputValue := reflect.ValueOf(r.Output).Elem()
+
+	updateMetadataFromNumValue(responseValue, "StatusCode", "status_code", res.Metadata)
+	res.Metadata["region"] = r.Region
+
+	if metadataOnly {
+		return
+	}
+
 	updateMetadataFromValue(inputValue, "Key", "key", res.Metadata)
-	outputValue := reflect.ValueOf(r.Data).Elem()
+	res.Metadata["request_id"] = r.RequestID
+
 	etag, ok := getFieldStringPtr(outputValue, "ETag")
 	if ok {
 		etag = strings.Trim(etag, "\"")
 		res.Metadata["etag"] = etag
 	}
 }
 
+
 func handleS3GetOrHeadObject(
-	r *aws.Request,
+	r *AWSCall,
 	res *protocol.Resource,
 	metadataOnly bool,
 	_ tracer.Tracer,
 ) {
-	commonS3OpertionHandler(r, res, metadataOnly)
-	outputValue := reflect.ValueOf(r.Data).Elem()
-	updateMetadataFromValue(outputValue, "ContentLength", "file_size", res.Metadata)
+	commonS3OperationHandler(r, res, metadataOnly)
 
-	lastModifiedField := outputValue.FieldByName("LastModified")
-	if lastModifiedField == (reflect.Value{}) {
-		return
-	}
-	lastModified := lastModifiedField.Elem().Interface().(time.Time)
-	res.Metadata["last_modified"] = lastModified.String()
+	outputValue := reflect.ValueOf(r.Output).Elem()
+	updateMetadataFromNumValue(outputValue, "ContentLength", "size", res.Metadata)
+	updateMetadataFromNumValue(outputValue, "LastModified", "last_modified", res.Metadata)
 }
 
 func handleS3PutObject(
-	r *aws.Request,
+	r *AWSCall,
 	res *protocol.Resource,
 	metadataOnly bool,
 	_ tracer.Tracer,
 ) {
-	commonS3OpertionHandler(r, res, metadataOnly)
+	commonS3OperationHandler(r, res, metadataOnly)
+
+	outputValue := reflect.ValueOf(r.Output).Elem()
+	updateMetadataFromNumValue(outputValue, "ContentLength", "size", res.Metadata)
 }
 
 type s3File struct {
-	key  string
-	size int64
-	etag string
+	key  string	`json:"key"`
```

Comment:
are the json tags needed?

Reply:
they dont hurt. they only help on Marshalling